packages/flame/lib/src/experimental/geometry/shapes/rounded_rectangle.dart

```diff
@@ -0,0 +1,182 @@
+import 'dart:math';
+import 'dart:ui';
+
+import 'package:flutter/cupertino.dart';
+import 'package:meta/meta.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+import '../../../game/transform2d.dart';
+import 'shape.dart';
+
+/// An axis-aligned rectangle with rounded corners.
+///
+/// The rounded parts of the rectangle are symmetrical in x- and y-directions,
+/// and across all corners.
+class RoundedRectangle extends Shape {
+  /// Constructs a [RoundedRectangle] with left, top, right and bottom edges,
+  /// and the given radius.
+  ///
+  /// If the edges are given in the wrong order (e.g. `left` is to the right
+  /// from `right`), then they will be swapped.
+  ///
+  /// The radius cannot be negative. At the same time, if the radius is too big,
+  /// it will be reduced so that the rounded edge can fit inside the rectangle.
+  /// In other words, the radius will be adjusted to not exceed the half-width
+  /// or half-height of the rectangle.
+  RoundedRectangle.fromLTRBR(
+    this._left,
+    this._top,
+    this._right,
+    this._bottom,
+    this._radius,
+  ) : assert(_radius >= 0, 'Radius cannot be negative: $_radius') {
+    if (_left > _right) {
+      final tmp = _left;
+      _left = _right;
+      _right = tmp;
+    }
+    if (_top > _bottom) {
+      final tmp = _top;
+      _top = _bottom;
+      _bottom = tmp;
+    }
+    if (_radius > (_right - _left) / 2) {
+      _radius = (_right - _left) / 2;
+    }
+    if (_radius > (_bottom - _top) / 2) {
+      _radius = (_bottom - _top) / 2;
+    }
+  }
+
+  factory RoundedRectangle.fromPoints(Vector2 a, Vector2 b, double radius) =>
+      RoundedRectangle.fromLTRBR(a.x, a.y, b.x, b.y, radius);
+
+  /// Constructs a [RoundedRectangle] from ui's [RRect].
+  ///
+  /// All corners of the `rrect` must have the same circular radii.
+  factory RoundedRectangle.fromRRect(RRect rrect) {
+    final radius = rrect.brRadiusX;
+    assert(
+      rrect.blRadiusX == radius &&
+          rrect.brRadiusX == radius &&
+          rrect.tlRadiusX == radius &&
+          rrect.trRadiusX == radius &&
+          rrect.blRadiusY == radius &&
+          rrect.brRadiusY == radius &&
+          rrect.tlRadiusY == radius &&
+          rrect.trRadiusY == radius,
+      'Unequal radii in the $rrect',
+    );
+    return RoundedRectangle.fromLTRBR(
+      rrect.left,
+      rrect.top,
+      rrect.right,
+      rrect.bottom,
+      radius,
+    );
+  }
+
+  double _left;
+  double _top;
+  double _right;
+  double _bottom;
+  double _radius;
+
+  double get left => _left;
+  double get right => _right;
+  double get top => _top;
+  double get bottom => _bottom;
+  double get radius => _radius;
+  double get width => _right - _left;
+  double get height => _bottom - _top;
+
+  @override
+  bool get isConvex => true;
+
+  @override
+  double get perimeter => (tau - 8) * _radius + 2 * (width + height);
```

Comment:
Why is `tao - 8` used here?

Reply:
Imagine you cut off the 4 rounded corners. Since each corner is a quarter of a circle, you can make a full circle out of them, with perimeter tau*r. The remaining sides each is shorter by 2r now. So their perimeter is (width-2r + height-2r)*2, which is (width+height)*2 - 8r. Adding back the circle, you have (tau-8)r.

Reply:
Makes sense, thanks for the explanation!